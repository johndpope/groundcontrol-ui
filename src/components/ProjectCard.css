.ProjectCard .ui.label {
  margin: .6em .8em 0 0;
}

.ProjectCard.ui.card > .content > .description {
  margin: 1em 0;
}

.ProjectCard .ui.menu {
  min-height: 2.7em;
}
