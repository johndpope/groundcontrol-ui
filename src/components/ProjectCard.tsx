// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import graphql from "babel-plugin-relay/macro";
import React from "react";
import { createFragmentContainer } from "react-relay";
import {
  Button,
  Card,
  Dimmer,
  Label,
  Loader,
  Tab,
 } from "semantic-ui-react";

import { ProjectCard_item } from "./__generated__/ProjectCard_item.graphql";

import CommitFeed from "./CommitFeed";
import RepositoryShortName from "./RepositoryShortName";

import "./ProjectCard.css";

export interface IProps {
  item: ProjectCard_item;
  onClone: (values: IProps) => any;
  onClickPath: (values: IProps) => any;
  onPull: (values: IProps) => any;
}

export function ProjectCard(props: IProps) {
  const {
    item,
    item: {
      repository,
      remoteReferenceShort,
      description,
      path,
      shortPath,
      remoteCommits,
      isCloned,
      isCloning,
      isPulling,
      isAhead,
      isBehind,
      isClean,
      localReferenceShort,
    },
    onClickPath,
    onClone,
    onPull,
  } = props;
  const remoteCommitNodes = item.remoteCommits.edges.map(({ node }) => node);
  const localCommitNodes = item.localCommits.edges.map(({ node }) => node);
  const handleClickPath = () => onClickPath({ ...props });
  const handleClone = () => onClone({ ...props });
  const handlePull = () => onPull({ ...props });
  const reference = remoteReferenceShort === localReferenceShort ?
    remoteReferenceShort : `${localReferenceShort} Â» ${remoteReferenceShort}`;
  const labels: JSX.Element[] = [];
  const buttons: JSX.Element[] = [];
  const menuProps = {
    size: "tiny",
    tabular: true,
  };
  const panes = [{
    menuItem: "Remote",
    render: () => <CommitFeed items={remoteCommitNodes} />,
  }];

  if (isCloned) {
    labels.push((
      <Label
        key="cloned"
        content="cloned"
        color="blue"
        size="small"
      />
    ));

    buttons.push((
      <Button
        key="pull"
        content="Pull"
        color="teal"
        icon="download"
        disabled={isPulling || !isBehind}
        loading={isPulling}
        onClick={handlePull}
      />
    ));

    if (!isBehind && !isAhead) {
      labels.push((
        <Label
          key="uptodate"
          content="up-to-date"
          color="teal"
          size="small"
        />
      ));
    }

    panes.push({
      menuItem: "Local",
      render: () => <CommitFeed items={localCommitNodes} />,
    });
  } else {
    buttons.push((
      <Button
        key="clone"
        content="Clone"
        color="teal"
        icon="clone"
        disabled={isCloning}
        loading={isCloning}
        onClick={handleClone}
      />
    ));
  }

  if (isAhead) {
    labels.push((
      <Label
        key="ahead"
        content="ahead"
        color="violet"
        size="small"
      />
    ));
  }

  if (isBehind) {
    labels.push((
      <Label
        key="behind"
        content="behind"
        color="purple"
        size="small"
      />
    ));
  }

  if (!isClean) {
    labels.push((
      <Label
        key="dirty"
        content="dirty"
        color="pink"
        size="small"
      />
    ));
  }

  let meta: JSX.Element | null = null;

  if (isCloned) {
    meta = (
      <Card.Meta
        as="a"
        href={`file://${path}`}
        onClick={handleClickPath}
      >
        {shortPath}
      </Card.Meta>
    );
  }

  return (
    <Card className="ProjectCard">
      <Card.Content>
        <Card.Header>
          <RepositoryShortName repository={repository} />
        </Card.Header>
        {meta}
        <Label
          size="small"
          color="grey"
        >
          {reference}
        </Label>
        {labels}
        <Card.Description>
          {description || "No description."}
        </Card.Description>
        <Tab
          panes={panes}
          menu={menuProps}
        />
      </Card.Content>
      <Card.Content extra={true}>
        <div className="ui three buttons">
          {buttons}
        </div>
      </Card.Content>
      <Dimmer
        active={remoteCommits.edges.length < 1}
        inverted={true}
      >
        <Loader content="Loading project commits..." />
      </Dimmer>
    </Card>
  );
}

export default createFragmentContainer(ProjectCard, graphql`
  fragment ProjectCard_item on Project
    @argumentDefinitions(
      commitsLimit: { type: "Int", defaultValue: 3 },
    ) {
    id
    repository
    remoteReferenceShort
    localReferenceShort
    description
    path
    shortPath
    isCloning
    isCloned
    isPulling
    isBehind
    isAhead
    isClean
    remoteCommits(first: $commitsLimit) {
      edges {
        node {
          ...CommitFeed_items
        }
      }
    }
    localCommits(first: $commitsLimit) {
      edges {
        node {
          ...CommitFeed_items
        }
      }
    }
  }`,
);
