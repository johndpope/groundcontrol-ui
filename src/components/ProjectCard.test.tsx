// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallow } from "enzyme";
import React from "react";
import { mocked } from "ts-jest/utils";

import { ProjectCard_item } from "./__generated__/ProjectCard_item.graphql";

import { mockQueryPropAttrs } from "../testing/relay";

import { IProps, ProjectCard } from "./ProjectCard";

jest.mock("./RepositoryShortName", () => "RepositoryShortName");

const item: ProjectCard_item = {
  ...mockQueryPropAttrs(),
  description: "description",
  id: "id",
  isAhead: false,
  isBehind: false,
  isClean: true,
  isCloned: false,
  isCloning: false,
  isPulling: false,
  localCommits: {
    edges: [{
      node: {
        ...mockQueryPropAttrs(),
      },
    }, {
      node: {
        ...mockQueryPropAttrs(),
      },
    }],
  },
  localReferenceShort: "localReferenceShort",
  path: "/long/path",
  remoteCommits: {
    edges: [{
      node: {
        ...mockQueryPropAttrs(),
      },
    }, {
      node: {
        ...mockQueryPropAttrs(),
      },
    }],
  },
  remoteReferenceShort: "remoteReferenceShort",
  repository: "repository",
  shortPath: "short/path",
};

const props = {
  item,
  onClickPath: jest.fn(),
  onClone: jest.fn(),
  onPull: jest.fn(),
};

beforeEach(() => {
  mocked(props.onClickPath).mockClear();
  mocked(props.onClone).mockClear();
  mocked(props.onPull).mockClear();
});

describe("<ProjectCard />", () => {

  it("renders correctly when not cloned", () => {
    const wrapper = shallow(
      <ProjectCard {...{...props, item: {...props.item, isCloned: false}}} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly when cloned", () => {
    const wrapper = shallow(
      <ProjectCard {...{...props, item: {...props.item, isCloned: true}}} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("shows a clone button if the project isn't cloned", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: false, isCloning: false } }}
      />,
    );
    expect(wrapper.find("Button[icon='clone'][loading=false]")).toHaveLength(1);
  });

  it("shows a loading clone button if the project is cloning", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: false, isCloning: true } }}
      />,
    );
    expect(wrapper.find("Button[icon='clone'][loading=true]")).toHaveLength(1);
  });

  it("shows a pull button if the project is cloned and behind", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isBehind: true } }}
      />,
    );
    expect(wrapper.find("Button[icon='download'][disabled=false]")).toHaveLength(1);
  });

  it("shows a disabled pull button if the project is cloned and up-to-date", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isBehind: false } }}
      />,
    );
    expect(wrapper.find("Button[icon='download'][disabled=true]")).toHaveLength(1);
  });

  it("shows a loading pull button if the project is pulling", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isPulling: true } }}
      />,
    );
    expect(wrapper.find("Button[icon='download'][loading=true]")).toHaveLength(1);
  });

  it("shows a cloned label if the project is cloned", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true } }}
      />,
    );
    expect(wrapper.find("Label[content='cloned']")).toHaveLength(1);
  });

  it("shows an up-to-date label if the project is up-to-date", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isBehind: false, isAhead: false } }}
      />,
    );
    expect(wrapper.find("Label[content='up-to-date']")).toHaveLength(1);
  });

  it("shows a behind label if the project is behind", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isBehind: true } }}
      />,
    );
    expect(wrapper.find("Label[content='behind']")).toHaveLength(1);
  });

  it("shows an ahead label if the project is ahead", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isAhead: true } }}
      />,
    );
    expect(wrapper.find("Label[content='ahead']")).toHaveLength(1);
  });

  it("shows a dirty label if the project isn't clean", () => {
    const wrapper = shallow(
      <ProjectCard
        {...{...props, item: { ...item, isCloned: true, isClean: false } }}
      />,
    );
    expect(wrapper.find("Label[content='dirty']")).toHaveLength(1);
  });

  it("triggers onClickPath when the path is clicked", () => {
    const wrapper = shallow(<ProjectCard {...props} />);
    wrapper.find("[href='file:///long/path']").simulate("click");
    expect(props.onClickPath).toBeCalledWith(props);
  });

  it("triggers onClone when the clone button is clicked", () => {
    const cloneProps: IProps = {...props, item: { ...item, isCloned: false } };
    const wrapper = shallow(<ProjectCard {...cloneProps} />);
    wrapper.find("Button[icon='clone']").simulate("click");
    expect(cloneProps.onClone).toBeCalledTimes(1);
    expect(cloneProps.onClone).toBeCalledWith(cloneProps);
  });

  it("triggers onPull when the pull button is clicked", () => {
    const pullProps: IProps = {...props, item: { ...item, isCloned: true, isBehind: true } };
    const wrapper = shallow(<ProjectCard {...pullProps} />);
    wrapper.find("Button[icon='download']").simulate("click");
    expect(pullProps.onPull).toBeCalledTimes(1);
    expect(pullProps.onPull).toBeCalledWith(pullProps);
  });

});
