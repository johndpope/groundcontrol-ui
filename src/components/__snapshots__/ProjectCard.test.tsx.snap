// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectCard /> renders correctly when cloned 1`] = `
<Card
  className="ProjectCard"
  color="teal"
>
  <CardContent>
    <CardHeader>
      <RepositoryShortName
        repository="repository"
      />
    </CardHeader>
    <Label
      size="small"
    >
      localReferenceShort » remoteReferenceShort
    </Label>
    <Label
      color="blue"
      content="cloned"
      key="cloned"
      size="small"
    />
    <Label
      color="teal"
      content="up-to-date"
      key="uptodate"
      size="small"
    />
    <CardDescription>
      description
    </CardDescription>
    <Tab
      grid={
        Object {
          "paneWidth": 12,
          "tabWidth": 4,
        }
      }
      menu={
        Object {
          "size": "tiny",
          "tabular": true,
        }
      }
      panes={
        Array [
          Object {
            "menuItem": "Remote",
            "render": [Function],
          },
          Object {
            "menuItem": "Local",
            "render": [Function],
          },
        ]
      }
      renderActiveOnly={true}
    />
  </CardContent>
  <CardContent
    extra={true}
  >
    <div
      className="ui three buttons"
    >
      <Button
        as="button"
        color="teal"
        content="Pull"
        disabled={true}
        icon="download"
        key="pull"
        loading={false}
        onClick={[Function]}
      />
    </div>
  </CardContent>
  <Dimmer
    active={false}
    inverted={true}
  >
    <Loader
      content="Loading project commits..."
    />
  </Dimmer>
</Card>
`;

exports[`<ProjectCard /> renders correctly when not cloned 1`] = `
<Card
  className="ProjectCard"
  color="grey"
>
  <CardContent>
    <CardHeader>
      <RepositoryShortName
        repository="repository"
      />
    </CardHeader>
    <Label
      size="small"
    >
      localReferenceShort » remoteReferenceShort
    </Label>
    <CardDescription>
      description
    </CardDescription>
    <Tab
      grid={
        Object {
          "paneWidth": 12,
          "tabWidth": 4,
        }
      }
      menu={
        Object {
          "size": "tiny",
          "tabular": true,
        }
      }
      panes={
        Array [
          Object {
            "menuItem": "Remote",
            "render": [Function],
          },
        ]
      }
      renderActiveOnly={true}
    />
  </CardContent>
  <CardContent
    extra={true}
  >
    <div
      className="ui three buttons"
    >
      <Button
        as="button"
        color="teal"
        content="Clone"
        disabled={false}
        icon="clone"
        key="clone"
        loading={false}
        onClick={[Function]}
      />
    </div>
  </CardContent>
  <Dimmer
    active={false}
    inverted={true}
  >
    <Loader
      content="Loading project commits..."
    />
  </Dimmer>
</Card>
`;
